libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,729 @@
+	// for every index in the mask copy out the indexed value in the source
+	switch (t)
+	{
+	case 0:
+		for (int i = 0; i < n; ++i)
+			v8x32 (m_SPP[0])[i] = v8x32(m_SP[1]) [v8x32 (m_SP[0])[i] % n];
+		break;
+	case 1:
+		for (int i = 0; i < n; ++i)
+			v16x16(m_SPP[0])[i] = v16x16(m_SP[1])[v16x16(m_SP[0])[i] % n];
+		break;
+	case 2:
+		for (int i = 0; i < n; ++i)
+			v32x8 (m_SPP[0])[i] = v32x8(m_SP[1]) [v32x8 (m_SP[0])[i] %  n];
+		break;
+	case 3:
+		for (int i = 0; i < n; ++i)
+			v64x4 (m_SPP[0])[i] = v64x4(m_SP[1]) [v64x4 (m_SP[0])[i] %  n];
+		break;
+	default:
+		throwBadInstruction();
+	}
+}
+
+void VM::xshuffle(uint8_t _b)
+{
+	// n type t elements in source and mask vectors
+	uint8_t const n = nElem(_b);
+	uint8_t const t = elemType(_b);
+
+	// given the type of the sources and mask                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
+	// for every index in the mask copy out the indexed value in one of the sources
+	switch (t)
+	{
+	case 0:
+		for (int i = 0; i < n; ++i)
+		{
+			int j = v8x32(m_SP[0]) [i];
+			v8x32 (m_SPP[0])[i] = j < n ? v8x32(m_SP[2]) [j] : v8x32 (m_SP[2])[j % n];
```

Comment:
This doesn't access source 1 vector, should be `j < n ? v8x32(m_SP[1]) [j] : v8x32 (m_SP[2])[j % n];` as I understand